src/github.com/matrix-org/dendron/dendron/main.go

```diff
@@ -148,49 +125,67 @@ func main() {
 		}
 	}
 
-	var synapse *exec.Cmd
-	var pusher *exec.Cmd
-	var synchrotron *exec.Cmd
-
 	var synapseLog = log.WithField("synapse", synapseURL.String())
 
+	// Used to terminate dendron.
+	terminate := make(chan string, 1)
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		s := <-signals
+		terminate <- fmt.Sprint("Got signal", s)
+	}()
+
 	if *startSynapse {
-		synapse = exec.Command(*synapsePython, "-m", "synapse.app.homeserver", "-c", *synapseConfig)
+		synapse := exec.Command(*synapsePython, "-m", "synapse.app.homeserver", "-c", *synapseConfig)
 		synapse.Stderr = os.Stderr
 		synapseLog.Print("Starting synapse")
 
 		synapse.Start()
+		defer synapse.Process.Signal(os.Interrupt)
 
-		sh := signalHander{
-			synapseLog: synapseLog,
-			synapse:    synapse.Process,
-		}
-		channel := make(chan os.Signal, 1)
-		signal.Notify(channel, os.Interrupt)
-		go sh.handleSignal(channel)
-
+		// Wait for synapse to start.
 		if err := waitForSynapse(synapseURL, synapseLog); err != nil {
 			synapseLog.Panic(err)
 		}
 
+		go func() {
+			// Wait for synapse to stop.
+			synapse.Process.Wait()
```

Comment:
It's a shame we're losing useful information like exit codes here. https://golang.org/pkg/os/#ProcessState and https://golang.org/pkg/syscall/#WaitStatus